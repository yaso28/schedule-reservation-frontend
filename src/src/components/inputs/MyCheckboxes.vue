<template>
  <MyFormGroup
    :label="label"
    :form="form"
    :field="field"
  >
    <MyCheckboxesNoWrapper
      :form="form"
      :field="field"
      :options="options"
      :value-mapper="valueMapper"
      :label-mapper="labelMapper"
      :inline="inline"
    >
      <template #label="{option}">
        <slot
          name="label"
          :option="option"
        />
      </template>
    </MyCheckboxesNoWrapper>
  </MyFormGroup>
</template>

<script>
export default {
  name: 'MyCheckboxes',
  components: {
    MyFormGroup: require('./MyFormGroup').default,
    MyCheckboxesNoWrapper: require('./MyCheckboxesNoWrapper').default
  },
  mixins: [require('./input-mixin').default],
  props: {
    options: {
      type: Array,
      required: true
    },
    valueMapper: {
      type: Function,
      default: o => o.id
    },
    labelMapper: {
      type: Function,
      default: o => o.name
    },
    inline: {
      type: Boolean,
      default: false
    }
  }
}
</script>
