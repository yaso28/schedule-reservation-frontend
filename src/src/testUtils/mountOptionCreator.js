import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
