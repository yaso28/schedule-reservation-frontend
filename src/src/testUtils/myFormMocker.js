import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
