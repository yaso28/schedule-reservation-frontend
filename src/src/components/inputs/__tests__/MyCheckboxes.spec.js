import { mount } from '@vue/test-utils'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import MyCheckboxes from '../MyCheckboxes'
import myFormMocker from '@/testUtils/myFormMocker'

jest.mock('@/models/MyForm')

function getInputWrapperArray (wrapper) {
  return wrapper.findAll('input')
}

const options = [
  { id: 1, name: 'オプション1' },
  { id: 2, name: 'オプション2' },
  { id: 3, name: 'オプション3' }
]
const valueBeforeCheckOn = []
const checkOnIndex = 1
const valueAfterCheckOn = [2]
const valueBeforeCheckOff = [1, 2, 3]
const checkOffIndex = 2
const valueAfterCheckOff = [1, 2]

describe('MyCheckboxes', () => {
  let form
  let field
  let mountOption
  let wrapper
  beforeEach(() => {
    const label = 'サンプル'
    field = 'sample_field'
    form = myFormMocker.createMockSingleField(field, [])

    mountOption = mountOptionCreator.create()
    mountOption.propsData = { label, form, field, options }
    wrapper = mount(MyCheckboxes, mountOption)
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('通常', () => {
    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('バリデーションエラー', () => {
    beforeEach(async () => {
      await myFormMocker.setError(wrapper, form, field, 'エラーです')
    })

    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('入力', () => {
    let inputWrapperArray
    beforeEach(() => {
      inputWrapperArray = getInputWrapperArray(wrapper)
    })

    describe('チェックON', () => {
      beforeEach(async () => {
        await myFormMocker.setValue(wrapper, form, field, valueBeforeCheckOn)
        await inputWrapperArray.at(checkOnIndex).setChecked(true)
      })

      test('MyForm.setValue()を呼出', () => {
        expect(form.setValue).toHaveBeenCalledWith(field, valueAfterCheckOn)
      })
    })

    describe('チェックOFF', () => {
      beforeEach(async () => {
        await myFormMocker.setValue(wrapper, form, field, valueBeforeCheckOff)
        await inputWrapperArray.at(checkOffIndex).setChecked(false)
      })

      test('MyForm.setValue()を呼出', () => {
        expect(form.setValue).toHaveBeenCalledWith(field, valueAfterCheckOff)
      })
    })
  })

  describe('オプション', () => {
    describe('custom mappers', () => {
      beforeEach(() => {
        const customOptions = [
          { key: 2, value: 'カスタム２' },
          { key: 4, value: 'カスタム４' }
        ]
        wrapper.setProps({
          options: customOptions,
          valueMapper: o => o.key,
          labelMapper: o => o.value
        })
      })

      test('表示', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })

    describe('inline', () => {
      beforeEach(() => {
        wrapper.setProps({
          inline: true
        })
      })

      test('表示', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('slot', () => {
    beforeEach(() => {
      mountOption.scopedSlots = {
        label (props) {
          return <span class="aaa">{props.option.name}</span>
        }
      }
      wrapper = mount(MyCheckboxes, mountOption)
    })

    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
