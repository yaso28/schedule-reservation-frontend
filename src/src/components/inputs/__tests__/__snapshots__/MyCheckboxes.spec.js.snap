// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyCheckboxes slot 表示 1`] = `
<div
  class="form-row form-group"
>
  <label
    class="col-form-label col-sm-3"
    for=""
  >
    
    サンプル
  
  </label>
   
  <div
    class="col-sm-9"
  >
    <div
      class="my-input-wrapper"
    >
      <div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-1"
            type="checkbox"
            value="1"
          />
           
          <label
            class="form-check-label"
            for="sample_field-1"
          >
            <span
              class="aaa"
            >
              オプション1
            </span>
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-2"
            type="checkbox"
            value="2"
          />
           
          <label
            class="form-check-label"
            for="sample_field-2"
          >
            <span
              class="aaa"
            >
              オプション2
            </span>
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-3"
            type="checkbox"
            value="3"
          />
           
          <label
            class="form-check-label"
            for="sample_field-3"
          >
            <span
              class="aaa"
            >
              オプション3
            </span>
          </label>
        </div>
      </div>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`MyCheckboxes オプション custom mappers 表示 1`] = `
<div
  class="form-row form-group"
>
  <label
    class="col-form-label col-sm-3"
    for=""
  >
    
    サンプル
  
  </label>
   
  <div
    class="col-sm-9"
  >
    <div
      class="my-input-wrapper"
    >
      <div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-2"
            type="checkbox"
            value="2"
          />
           
          <label
            class="form-check-label"
            for="sample_field-2"
          >
            
        カスタム２
      
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-4"
            type="checkbox"
            value="4"
          />
           
          <label
            class="form-check-label"
            for="sample_field-4"
          >
            
        カスタム４
      
          </label>
        </div>
      </div>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`MyCheckboxes オプション inline 表示 1`] = `
<div
  class="form-row form-group"
>
  <label
    class="col-form-label col-sm-3"
    for=""
  >
    
    サンプル
  
  </label>
   
  <div
    class="col-sm-9"
  >
    <div
      class="my-input-wrapper"
    >
      <div>
        <div
          class="form-check form-check-inline"
        >
          <input
            class="form-check-input"
            id="sample_field-1"
            type="checkbox"
            value="1"
          />
           
          <label
            class="form-check-label"
            for="sample_field-1"
          >
            
        オプション1
      
          </label>
        </div>
        <div
          class="form-check form-check-inline"
        >
          <input
            class="form-check-input"
            id="sample_field-2"
            type="checkbox"
            value="2"
          />
           
          <label
            class="form-check-label"
            for="sample_field-2"
          >
            
        オプション2
      
          </label>
        </div>
        <div
          class="form-check form-check-inline"
        >
          <input
            class="form-check-input"
            id="sample_field-3"
            type="checkbox"
            value="3"
          />
           
          <label
            class="form-check-label"
            for="sample_field-3"
          >
            
        オプション3
      
          </label>
        </div>
      </div>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`MyCheckboxes バリデーションエラー 表示 1`] = `
<div
  class="form-row form-group"
>
  <label
    class="col-form-label col-sm-3"
    for=""
  >
    
    サンプル
  
  </label>
   
  <div
    class="col-sm-9"
  >
    <div
      class="my-input-wrapper is-invalid"
    >
      <div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-1"
            type="checkbox"
            value="1"
          />
           
          <label
            class="form-check-label"
            for="sample_field-1"
          >
            
        オプション1
      
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-2"
            type="checkbox"
            value="2"
          />
           
          <label
            class="form-check-label"
            for="sample_field-2"
          >
            
        オプション2
      
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-3"
            type="checkbox"
            value="3"
          />
           
          <label
            class="form-check-label"
            for="sample_field-3"
          >
            
        オプション3
      
          </label>
        </div>
      </div>
    </div>
     
    <div
      class="invalid-feedback"
    >
      
      エラーです
    
    </div>
  </div>
</div>
`;

exports[`MyCheckboxes 通常 表示 1`] = `
<div
  class="form-row form-group"
>
  <label
    class="col-form-label col-sm-3"
    for=""
  >
    
    サンプル
  
  </label>
   
  <div
    class="col-sm-9"
  >
    <div
      class="my-input-wrapper"
    >
      <div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-1"
            type="checkbox"
            value="1"
          />
           
          <label
            class="form-check-label"
            for="sample_field-1"
          >
            
        オプション1
      
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-2"
            type="checkbox"
            value="2"
          />
           
          <label
            class="form-check-label"
            for="sample_field-2"
          >
            
        オプション2
      
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="sample_field-3"
            type="checkbox"
            value="3"
          />
           
          <label
            class="form-check-label"
            for="sample_field-3"
          >
            
        オプション3
      
          </label>
        </div>
      </div>
    </div>
     
    <!---->
  </div>
</div>
`;
